### PR TITLE
Implement rc curve generation in c++

### LINKED ISSUE #2787: Port RC Curve Generation to C++
Port the Python script that generates RC curves to C++.

This will be intended for non-consensus use only and will contain some floating point calculations to make porting easier.

### PR DESCRIPTION
**Pull request by mvandeberg**
Closes #2787

Due to my decision to keep this as a local header (to prevent accidental inclusion in consensus code) testing is difficult. Below are the results of the existing curves from the python and c++ implementations.

Python Reference Implementation:
```
{
 "resource_parameters":[
  [
   "resource_history_bytes",
   [
    {
     "time_unit":"rc_time_unit_blocks",
     "resource_unit":1,
     "budget_per_time_unit":34722,
     "pool_eq":21640431401,
     "max_pool_size":"43280862801",
     "curve_params":{
      "coeff_a":"9736235291562360832",
      "coeff_b":"169065870",
      "coeff_d":"0",
      "shift":47
     },
     "decay_params":{
      "decay_per_time_unit":"3613026481",
      "decay_per_time_unit_denom_shift":51
     }
    },
    {}
   ]
  ],
[
   "resource_new_accounts",
   [
    {
     "time_unit":"rc_time_unit_blocks",
     "resource_unit":10000,
     "budget_per_time_unit":0,
     "pool_eq":14426950409,
     "max_pool_size":"28853900818",
     "curve_params":{
      "coeff_a":"9904210705121218560",
      "coeff_b":"112710550",
      "coeff_d":"0",
      "shift":31
     },
     "decay_params":{
      "decay_per_time_unit":"3551752401",
      "decay_per_time_unit_denom_shift":67
     }
    },
    {}
   ]
  ],
  [
   "resource_market_bytes",
   [
    {
     "time_unit":"rc_time_unit_blocks",
     "resource_unit":100,
     "budget_per_time_unit":578704,
     "pool_eq":16030041350,
     "max_pool_size":"32060082699",
     "curve_params":{
      "coeff_a":"13847090192444250112",
      "coeff_b":"125234698",
      "coeff_d":"0",
      "shift":52
     },
     "decay_params":{
      "decay_per_time_unit":"2540365427",
      "decay_per_time_unit_denom_shift":46
     }
    },
    {}
   ]
  ],
  [
   "resource_state_bytes",
   [
    {
     "time_unit":"rc_time_unit_blocks",
     "resource_unit":1,
     "budget_per_time_unit":23148148,
     "pool_eq":14426954266915,
     "max_pool_size":"28853908533829",
     "curve_params":{
      "coeff_a":"9736235291562360832",
      "coeff_b":"112710580210",
      "coeff_d":"0",
      "shift":47
     },
     "decay_params":{
      "decay_per_time_unit":"3613026481",
      "decay_per_time_unit_denom_shift":51
     }
    },
    {}
   ]
  ],
  [
   "resource_execution_time",
   [
    {
     "time_unit":"rc_time_unit_seconds",
     "resource_unit":1,
     "budget_per_time_unit":2739726,
     "pool_eq":5122556912307,
     "max_pool_size":"10245113824613",
     "curve_params":{
      "coeff_a":"9736235291562360832",
      "coeff_b":"40019975877",
      "coeff_d":"0",
      "shift":47
     },
     "decay_params":{
      "decay_per_time_unit":"2408685609",
      "decay_per_time_unit_denom_shift":52
     }
    },
    {}
   ]
  ]
 ]
}
```

C++ Implementation
```
[
  {
    "time_unit": 1,
    "resource_unit": 1,
    "budget_per_time_unit": 34722,
    "pool_eq": "21640431401",
    "max_pool_size": "43280862801",
    "curve_params": {
      "coeff_a": "9736235291562395648",
      "coeff_b": 169065870,
      "shift": 47
    },
    "decay_params": {
      "decay_per_time_unit": 3613026481,
      "decay_per_time_unit_denom_shift": 51
    }
  },
  {
    "time_unit": 1,
    "resource_unit": 10000,
    "budget_per_time_unit": 0,
    "pool_eq": "14426950409",
    "max_pool_size": "28853900818",
    "curve_params": {
      "coeff_a": "9904210707427221504",
      "coeff_b": 112710550,
      "shift": 31
    },
    "decay_params": {
      "decay_per_time_unit": 3551752401,
      "decay_per_time_unit_denom_shift": 67
    }
  },
  {
    "time_unit": 1,
    "resource_unit": 100,
    "budget_per_time_unit": 578704,
    "pool_eq": "16030041350",
    "max_pool_size": "32060082699",
    "curve_params": {
      "coeff_a": "13847090192444250112",
      "coeff_b": 125234698,
      "shift": 52
    },
    "decay_params": {
      "decay_per_time_unit": 2540365427,
      "decay_per_time_unit_denom_shift": 46
    }
  },
  {
    "time_unit": 1,
    "resource_unit": 1,
    "budget_per_time_unit": 23148148,
    "pool_eq": "14426954266915",
    "max_pool_size": "28853908533829",
    "curve_params": {
      "coeff_a": "9736235291562395648",
      "coeff_b": "112710580210",
      "shift": 47
    },
    "decay_params": {
      "decay_per_time_unit": 3613026481,
      "decay_per_time_unit_denom_shift": 51
    }
  },
  {
    "time_unit": 0,
    "resource_unit": 1,
    "budget_per_time_unit": 2739726,
    "pool_eq": "5122556912307",
    "max_pool_size": "10245113824613",
    "curve_params": {
      "coeff_a": "9736235291562395648",
      "coeff_b": "40019975877",
      "shift": 47
    },
    "decay_params": {
      "decay_per_time_unit": 2408685609,
      "decay_per_time_unit_denom_shift": 52
    }
  }
]
```

### CODE REVIEW

File: libraries/plugins/rc/rc_curve.hpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+struct rc_curve_gen_params
+{
+   rc_time_unit_type       time_unit = rc_time_unit_seconds;
+   rc_time_type            budget_time;
+   uint64_t                budget;
+   rc_time_type            half_life;
+   uint32_t                inelasticity_threshold_num    = 1;
+   uint32_t                inelasticity_threshold_denom  = 128;
+   uint64_t                small_stockpile_size          = 0x0000000100000000ull; // 2^32
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.hpp]: i believe you can just write (1 << 32), or at least omit the leading zeroes. that constant isn't readable to human eyes

File: libraries/plugins/rc/rc_curve.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+         + fc::days( weeks * 7 )
+         + fc::days( months * 30 )
+         + fc::days( years * 365 );
+   }
+};
+
+struct rc_curve_gen_params
+{
+   rc_time_unit_type       time_unit = rc_time_unit_seconds;
+   rc_time_type            budget_time;
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.hpp]: give these default values, even if they're invalid ones. mixing defaulting and non-defaulting fields risks accesses to uninitialized memory

File: libraries/plugins/rc/rc_curve.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+{
+   try
+   {
+      const static double log_2 = std::log( 2.0 );
+
+      rc_resource_params result;
+
+      result.time_unit = params.time_unit;
+
+      uint32_t time_unit_sec = params.time_unit == rc_time_unit_seconds ? 1 : STEEM_BLOCK_INTERVAL;
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.cpp]: unclear, rewrite avoiding ternary operator or add comment

File: libraries/plugins/rc/rc_curve.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      double u_point = (double)( params.u_point_num ) / (double)( params.u_point_denom );
+      double k = u_point / ( a_point * ( 1.0 - u_point ) );
+
+      double B = inelasticity_threshold * pool_eq;
+      double A = tau / k;
+
+      if( A < 1.0 || B < 1.0 )
+      {
+         wlog( "Bad parameter vlaue (is time too short?)" );
+         FC_ASSERT( false, "Bad parameter vlaue (is time too short?)" );
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.cpp]: `value`

File: libraries/plugins/rc/rc_curve.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      }
+
+      double curve_shift_float = std::log( (0xFFFFFFFFFFFFFFFFull) / A ) / log_2;
```
Comment (goldibex) [inline libraries/plugins/rc/rc_curve.cpp]: prefer `(uint64_t)UINT64_MAX`?

File: libraries/plugins/rc/rc_curve.hpp (hunk 2 above)
Comment (mvandeberg) [reply libraries/plugins/rc/rc_curve.hpp]: `budget_time` is default initialized above (as `rc_time_type`). This is a convoluted way to define time, but it is intended to be json compatible with the python script if we were deserialize the same input file.

**Comment by goldibex** [comment]
@mvandeberg i'm with you on the rc_time_type initialization, so there's  just that change and the constant declaration left and this lgtm. @theoreticalbts have you had a chance to check the curve generation math yet?

**Comment by mvandeberg** [comment]
Obsolete by 2839

**Pull request closed without merging**